toolchain/check/handle_struct.cpp

```diff
@@ -52,6 +52,64 @@ auto HandleStructFieldValue(Context& context, Parse::Node parse_node) -> bool {
   return true;
 }
 
+// Goes from the StructFieldValue node pointing to the `=` in a struct literal
+// like `{.a = 1}`, to the name node `a` in its children.
+static auto StructFieldValueToName(Context& context,
+                                   Parse::Node field_value_node)
+    -> Parse::Node {
+  // TODO: Will be easier after #3393.
+  CARBON_CHECK(context.parse_tree().node_kind(field_value_node) ==
+               Parse::NodeKind::StructFieldValue);
+  auto children = context.parse_tree().children(field_value_node);
+  auto struct_field_designator_iter = children.begin();
+  ++struct_field_designator_iter;
+  CARBON_CHECK(context.parse_tree().node_kind(*struct_field_designator_iter) ==
+               Parse::NodeKind::StructFieldDesignator);
+  children = context.parse_tree().children(*struct_field_designator_iter);
+  return *children.begin();
+}
+
+static auto HasDuplicateNamesError(Context& context,
+                                   SemIR::InstBlockId type_block_id) -> bool {
+  auto& sem_ir = context.sem_ir();
+  auto fields = sem_ir.inst_blocks().Get(type_block_id);
+  llvm::DenseMap<SemIR::NameId, Parse::Node> names;
+  auto& insts = sem_ir.insts();
+  for (SemIR::InstId field_inst_id : fields) {
+    auto field_inst = insts.GetAs<SemIR::StructTypeField>(field_inst_id);
+    auto [it, added] =
+        names.insert({field_inst.name_id, field_inst.parse_node});
+    if (!added) {
+      // FIXME: Should this be renamed `MemberNameDuplicate`, and possibly
+      // made more widely available?
+      CARBON_DIAGNOSTIC(StructNameDuplicate, Error,
+                        "Member of {0} with a duplicated name `{1}`.",
+                        llvm::StringRef, llvm::StringRef);
+      CARBON_DIAGNOSTIC(StructNamePrevious, Note,
+                        "Member with the same name here.");
+      llvm::StringRef container;
+      Parse::Node prev_node = it->second;
+      if (context.parse_tree().node_kind(field_inst.parse_node) ==
+          Parse::NodeKind::StructFieldValue) {
+        container = "struct literal";
+        // This avoids using the wrong parse node in the struct literal case.
+        prev_node = StructFieldValueToName(context, prev_node);
+        field_inst.parse_node =
+            StructFieldValueToName(context, field_inst.parse_node);
+      } else {
```

Comment:
I think it would be better if the parse node associated with a `StructTypeField` were consistent. Right now, within a `StructTypeLiteral`, it points at the name, whereas in a `StructLiteral` it points at the `=`. So perhaps instead of walking from the `StructFieldValue` to the `Name` here, we could collect and store the name node directly in `HandleStructFieldValue`:

```diff
-  SemIR::NameId name_id = context.node_stack().Pop<Parse::NodeKind::Name>();
+  auto [name_node, name_id] =
+      context.node_stack().PopWithParseNode<Parse::NodeKind::Name>();
   // Store the name for the type.
   context.args_type_info_stack().AddInst(SemIR::StructTypeField{
-      parse_node, name_id, context.insts().Get(value_inst_id).type_id()});
+      name_node, name_id, context.insts().Get(value_inst_id).type_id()});
```

Reply:
What is going on here is that when there is a struct literal, it is deducing its type, and so the elements of that type are only loosely associated with the source code as written. It is fine to do what you are saying for now, but it is something of a hack relying on nothing else at the moment trying to use this data.